### ISSUE TITLE
"No Email Found" for Open Collective but does exist

### ISSUE THREAD

**Issue by tangjeff0** [athensresearch/athens#694]
At least 5 separate users have reported this.

The current implementation:
1. User submits their email in Athens
2. This email hits an Athens AWS Lambda
3. The Lambda hits OpenCollective's GraphQL v2 endpoint, with Api-Key authorization.
4. Doesn't appear as if we can find user's by email, so the GraphQL grabs ALL users, then filters by email
5. If the final data has > 1 length, where an item has the user's email, return success
6. User's UI is updated.

--- 

- [McMatt](https://discord.com/channels/708122962422792194/714804822909517935/814121011003523113)
- [qqq](https://discord.com/channels/708122962422792194/714804822909517935/814363568480714813)
- J Martinez

**Comment by tangjeff0** [comment]
@befree would you happen to know why this is happening?

**Comment by tangjeff0** [comment]
@betree would you happen to know why this is happening here? Is there a better way to check if a user has sponsored us through OpenCollective?

**Comment by Mirtma** [comment]
@Ezzie also: [link](https://discord.com/channels/708122962422792194/802652338619285594/814486113921794059)

**Comment by Betree** [comment]
The simplest solution I can think of would be:
- Listen to the `collective.member.created` event
- From the payload, get `member.memberCollective.slug`
- Run the following query:

```gql
{
  account(slug: "the-slug") {
    ... on Individual {
      email
    }
  }
}
```

If responding to an event webhook is not a good solution, the approach you've described here seems good.

As the query above implies, this only works for people contributing as individuals (not organizations). This might be the reason why the emails are not found. If not, I would need to know the profiles/emails to investigate further.

**Comment by shebang42** [comment]
> this only works for people contributing as individuals
I contribute as individual (since yesterday).

```
HTTP/1.1 200 OK
Content-Type: application/json
<..>

{"email_exists":false}
```

**Comment by Mirtma** [comment]
> this only works for people contributing as individuals

I've also contributed as an individual with the User profile.

**Comment by jmartinezheras** [comment]
same for me, I contributed as individual

![image](https://user-images.githubusercontent.com/38404513/109181493-457ed980-778c-11eb-9f53-57a4f5ba3910.png)

**Comment by Positron010** [comment]
> this only works for people contributing as individuals

Yes, it is the same for me. I also contributed as an individual.

**Comment by leogao2** [comment]
I contributed as an individual and am also receiving this error.

**Comment by xllora** [comment]
I get the same message for my account created under "community discount" for being part of the beta.

**Comment by tangjeff0** [comment]
@Betree there are only two organizations that have sponsored Athens. This is visible on our [Open Collective](https://opencollective.com/athens). I don't think either of them have tried validating their email yet through Athens. Either way, it's clear that more than 2 people have had this issue, and it seems to be mostly individuals.

I'm using email right now because eventually we can turn this into a Magic Link sign-in, which provides a fairly user-friendly and secure form of authentication for users.

Is there a better way for Athens to gain access to the emails of users that have sponsored Athens through Open Collective? Currently, we get an email for new members that includes the member's email when they sponsor Athens. We can also export the entire Membership CSV from our Open Collective. Is there a more programmatic way to validate users' emails?

**Comment by matlemai** [comment]
> I get the same message for my account created under "community discount" for being part of the beta.

Exactly the same for me as well.

**Comment by Betree** [comment]
> Is there a better way for Athens to gain access to the emails of users that have sponsored Athens through Open Collective? Currently, we get an email for new members that includes the member's email when they sponsor Athens. We can also export the entire Membership CSV from our Open Collective. Is there a more programmatic way to validate users' emails?

@tangjeff0 The approach I shared [in my comment above](https://github.com/athensresearch/athens/issues/694#issuecomment-785936612) (using the public GraphQL API) should be enough to get users emails given that the `Api-Key` provided in the headers has been generated from an account that has access to this information (in your case, the account of an admin of Athens).

So the following query should return all the emails of your members:

```gql
{
  account(slug: "athens") {
    members {
      nodes {
        id
        role
        account {
          id
          slug
          ... on Individual {
            email
          }
        }
      }
    }
  }
}
```

Something we can do to simplify this in the future (and make it faster) would be to add an `email` argument on `members`:

```diff gql
{
  account(slug: "athens") {
-   members {
+   members(email: "my-email@gmail.com") {
      nodes {
        id
        role
        account {
          id
          slug
          ... on Individual {
            email
          }
        }
      }
    }
  }
}
```

**Comment by tangjeff0** [comment]
Hey @Betree, we basically had the right query the entire time:

```
        account(slug: "athens") {
          id
          name
          slug
          members {
            nodes {
              id
              account {
                name
                type
                ... on Individual {
                  email
                }
                ... on Organization {
                  email
                }
              }
            }
          }
        }
```

I think what you said about `pagination` is most likely the reason why some of these requests are failing! We have to set the limit to be at least the entire number of members.

Will test this out soon but would love to have the email query in the meantime.

**Comment by tangjeff0** [comment]
Just pushed a fix which increases the `limit` to `1000`. I noticed by default the limit is `100`.

Assuming this works, querying the OpenCollective API by email directly will eventually by necessary for scalability. Otherwise, will probably have to migrate sooner than later off of OC.

**Comment by Mirtma** [comment]
> Just pushed a fix which increases the `limit` to `1000`. I noticed by default the limit is `100`.
> 
> Assuming this works, querying the OpenCollective API by email directly will eventually by necessary for scalability. Otherwise, will probably have to migrate sooner than later off of OC.

Yes, it's working now.

**Comment by xllora** [comment]
+1

**Comment by tangjeff0** [comment]
Seems to be fixed. Closing

**Comment by Correntze** [comment]
Hi Everyone, 
I am on the Windows desktop app and still received the error message (sponsored as part of the community discount).
Please let me know!
